src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageHandlerMethodSelector.cs

```diff
@@ -3,140 +3,106 @@
             }
 
-            var ambiguousMatches = (List<HandlerMethodDescriptor>)null;
-            var best = (HandlerMethodAndMetadata?)null;
-            for (var i = 2; i >= 0; i--)
+            List<HandlerMethodDescriptor> ambiguousMatches = null;
+            HandlerMethodDescriptor bestMatch = null;
+            for (var score = 2; score >= 0; score--)
             {
-                for (var j = 0; j < handlers.Count; j++)
+                for (var i = 0; i < handlers.Count; i++)
                 {
-                    var handler = handlers[j];
-                    if (handler.GetScore() == i)
+                    var handler = handlers[i];
+                    if (GetScore(handler) == score)
                     {
-                        if (best == null)
+                        if (bestMatch == null)
                         {
-                            best = handler;
+                            bestMatch = handler;
                             continue;
                         }
 
                         if (ambiguousMatches == null)
                         {
                             ambiguousMatches = new List<HandlerMethodDescriptor>();
-                            ambiguousMatches.Add(best.Value.Handler);
+                            ambiguousMatches.Add(bestMatch);
                         }
 
-                        ambiguousMatches.Add(handler.Handler);
+                        ambiguousMatches.Add(handler);
                     }
                 }
 
                 if (ambiguousMatches != null)
                 {
-                    throw new InvalidOperationException($"Selecting a handler is ambiguous! Matches: {string.Join(", ", ambiguousMatches)}");
+                    var ambiguousMethods = string.Join(", ", ambiguousMatches.Select(m => m.Method));
+                    throw new InvalidOperationException(Resources.FormatAmbiguousHandler(Environment.NewLine, ambiguousMethods));
                 }
 
-                if (best != null)
+                if (bestMatch != null)
                 {
-                    return best.Value.Handler;
+                    return bestMatch;
                 }
             }
 
             return null;
         }
 
-        // Bad prototype substring implementation :)
-        private struct HandlerMethodAndMetadata
+        private static List<HandlerMethodDescriptor> SelectHandlers(PageContext context)
         {
-            public static HandlerMethodAndMetadata Create(HandlerMethodDescriptor handler)
-            {
-                var name = handler.Method.Name;
+            var handlers = context.ActionDescriptor.HandlerMethods;
+            List<HandlerMethodDescriptor> handlersToConsider = null;
 
-                string httpMethod;
-                if (name.StartsWith("OnGet", StringComparison.Ordinal))
+            var formAction = Convert.ToString(context.RouteData.Values[FormAction]);
+            for (var i = 0; i < handlers.Count; i++)
```

Comment:
More of a question than feedback, you seem to prefer `for` over `foreach` even when you only use the `i` to get elements, is that just personal preference or something else?